nexus/src/app/silo.rs

```diff
@@ -0,0 +1,150 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Silos and Users.
+
+use crate::authz;
+use crate::context::OpContext;
+use crate::db;
+use crate::db::identity::Resource;
+use crate::db::lookup::LookupPath;
+use crate::db::model::Name;
+use crate::db::model::SshKey;
+use crate::external_api::params;
+use omicron_common::api::external::CreateResult;
+use omicron_common::api::external::DataPageParams;
+use omicron_common::api::external::DeleteResult;
+use omicron_common::api::external::ListResultVec;
+use omicron_common::api::external::LookupResult;
+use uuid::Uuid;
+
+impl super::Nexus {
+    // Silos
+
+    pub async fn silo_create(
+        &self,
+        opctx: &OpContext,
+        new_silo_params: params::SiloCreate,
+    ) -> CreateResult<db::model::Silo> {
+        let silo = db::model::Silo::new(new_silo_params);
+        self.db_datastore.silo_create(opctx, silo).await
+    }
+
+    pub async fn silos_list_by_name(
+        &self,
+        opctx: &OpContext,
+        pagparams: &DataPageParams<'_, Name>,
+    ) -> ListResultVec<db::model::Silo> {
+        self.db_datastore.silos_list_by_name(opctx, pagparams).await
+    }
+
+    pub async fn silos_list_by_id(
+        &self,
+        opctx: &OpContext,
+        pagparams: &DataPageParams<'_, Uuid>,
+    ) -> ListResultVec<db::model::Silo> {
+        self.db_datastore.silos_list_by_id(opctx, pagparams).await
+    }
+
+    pub async fn silo_fetch(
+        &self,
+        opctx: &OpContext,
+        name: &Name,
+    ) -> LookupResult<db::model::Silo> {
+        let (.., db_silo) = LookupPath::new(opctx, &self.db_datastore)
+            .silo_name(name)
+            .fetch()
+            .await?;
+        Ok(db_silo)
+    }
+
+    pub async fn silo_delete(
+        &self,
+        opctx: &OpContext,
+        name: &Name,
+    ) -> DeleteResult {
+        let (.., authz_silo, db_silo) =
+            LookupPath::new(opctx, &self.db_datastore)
+                .silo_name(name)
+                .fetch_for(authz::Action::Delete)
+                .await?;
+        self.db_datastore.silo_delete(opctx, &authz_silo, &db_silo).await
+    }
+
+    // Users
+
+    pub async fn silo_user_fetch(
+        &self,
+        opctx: &OpContext,
+        silo_user_id: Uuid,
+    ) -> LookupResult<db::model::SiloUser> {
+        let (.., db_silo_user) = LookupPath::new(opctx, &self.datastore())
+            .silo_user_id(silo_user_id)
+            .fetch()
+            .await?;
+        Ok(db_silo_user)
+    }
+
+    // SSH Keys
+
+    pub async fn ssh_key_create(
```

Comment:
A few things feel weird about this organization but I'm not sure what else to do.

silo_user has more in common with the other iam stuff (like user_builtin) than silos.
ssh keys don't really have anything to do with silos.  They _are_ related to silo users, so I guess they could all go into an "iam.rs" ... but ssh keys feels like stretching that.

Reply:
Grouping into `iam.rs` for now - this feels like a coarse-grained solution, but it does seem to be reasonably encapsulating. "Silos + users + ssh keys" seem much more tightly coupled to each other than any of them are to, say, a disk or an instance.